codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/ShapeVisitorHelper.java

```diff
@@ -140,17 +140,26 @@ public static String toDafnyShapeVisitorWriter(
     final Shape targetShape = context
       .model()
       .expectShape(memberShape.getTarget());
+    // Resource shape already goes into a function
     if (targetShape.hasTrait(ReferenceTrait.class)) {
-      return targetShape.accept(
-        new SmithyToDafnyShapeVisitor(
-          context,
-          dataSource,
-          writer,
-          isConfigShape,
-          isOptional,
-          isPointerType
-        )
+      final ReferenceTrait referenceTrait = targetShape.expectTrait(
+        ReferenceTrait.class
       );
+      final Shape resourceOrService = context
+        .model()
+        .expectShape(referenceTrait.getReferentId());
+      if (resourceOrService.isResourceShape()) {
```

Comment:
What happens if resource is a service type?

Reply:
We will do the same thing as rest of the shape. Put its implementation into a  function.